Today Extension/TodayViewController.swift

```diff
@@ -0,0 +1,226 @@
+    func widgetActiveDisplayModeDidChange(_ activeDisplayMode: NCWidgetDisplayMode, withMaximumSize maxSize: CGSize) {
+        let expanded = activeDisplayMode == .expanded
+        preferredContentSize = expanded ? CGSize(width: maxSize.width, height: 110.0 * CGFloat(favorites.count)) : maxSize
+
+        routesTable.reloadData()
+    }
+
+    func createConstraints() {
+        routesTable.snp.makeConstraints {(make) in
+            make.leading.trailing.equalToSuperview()
+            make.height.equalTo(110.0*5)
+        }
+    }
+
+    /// Re-orders routes so that walking-only routes are at the end of the list of routes. Modifies the list of favorites and coordinates coorespondingly.
+    func rearrangeRoutes() {
+        var nonNilRoutes = [Route?]()
+        var nonNilFavorites = [String]()
+        var nonNilCoordinates = [String]()
+
+        var nilRoutes = [Route?]()
+        var nilFavorites = [String]()
+        var nilCoordinates = [String]()
+
+        for i in 0..<routes.count {
+
+            if (routes[i]?.directions.filter { $0.type == .depart })?.first != nil {
+                nonNilRoutes.append(routes[i])
+                nonNilFavorites.append(favorites[i])
+                nonNilCoordinates.append(coordinates[i])
+            } else {
+                nilRoutes.append(routes[i])
+                nilFavorites.append(favorites[i])
+                nilCoordinates.append(coordinates[i])
+            }
+        }
+
+        favorites = nonNilFavorites + nilFavorites
+        coordinates = nonNilCoordinates + nilCoordinates
+        routes = nonNilRoutes + nilRoutes
+    }
+}
+
+extension TodayViewController: UITableViewDataSource, UITableViewDelegate {
+    private func setUpRoutesTableView() {
+        routesTable.delegate = self
+        routesTable.dataSource = self
+        routesTable.register(TodayExtensionCell.self, forCellReuseIdentifier: Constants.TodayExtension.contentCellIdentifier)
+        routesTable.register(TodayExtensionErrorCell.self, forCellReuseIdentifier: Constants.TodayExtension.errorCellIdentifier)
+        routesTable.register(LoadingTableViewCell.self, forCellReuseIdentifier: Constants.TodayExtension.loadingCellIdentifier)
+    }
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        let isCompact = extensionContext?.widgetActiveDisplayMode == .compact
+
+        if (isCompact || favorites.isEmpty || routes.isEmpty || !didFetchRoutes) {
+            // if: in compact mode, no favorites added, failed to load routes, or in loading state
+            return 1
+        } else {
+            return favorites.count
+        }
+
+    }
+
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return 110.0
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        if (favorites.count != 0) {
+            if (routes.isEmpty) {
```

Comment:
Can chain these two conditionals together

Reply:
Was messing around with this, but wasn't sure of the best way to chain the conditionals for the rest of the code. Within `favorites.count != 0`, there's cases for `routes.isEmpty` and `!routes.IsEmpty`, which is why I separated the two conditionals in the beginning.